File: threads/src/main/java/im/threads/ui/utils/FileHelper.java
package im.threads.ui.utils;

import im.threads.business.config.BaseConfig;
import im.threads.business.logger.LoggerEdna;
import im.threads.business.transport.models.AttachmentSettings;
import im.threads.internal.chat_updates.ChatUpdateProcessor;
import im.threads.ui.utils.preferences.PrefUtilsUi;

public enum FileHelper {

    INSTANCE;

    private static final double MEGABYTE = 1024 * 1024;

    FileHelper() {
        ChatUpdateProcessor.getInstance().getAttachmentSettingsProcessor()
                .subscribe(attachmentSettings -> saveAttachmentSettings(attachmentSettings.getContent()),
                        LoggerEdna::error
                );
    }

    public boolean isAllowedFileSize(long fileSize) {
        return fileSize / MEGABYTE <= getMaxAllowedFileSize();
    }

    public boolean isAllowedFileExtension(String fileExtension) {
        AttachmentSettings.Content attachmentSettings = getAttachmentSettings();
        for (String allowedExt : attachmentSettings.getFileExtensions()) {
            if (allowedExt.equalsIgnoreCase(fileExtension)) {
                return true;
            }
        }
        return false;
    }

    public long getMaxAllowedFileSize() {
        return getAttachmentSettings().getMaxSize();
    }

    private void saveAttachmentSettings(AttachmentSettings.Content attachmentSettingsContent) {
        PrefUtilsUi.setAttachmentSettings(BaseConfig.instance.gson.toJson(attachmentSettingsContent));
    }

    private AttachmentSettings.Content getAttachmentSettings() {
        String settingsStr = PrefUtilsUi.getAttachmentSettings();
        settingsStr = settingsStr == null ? "" : settingsStr;

        if (settingsStr.isEmpty()) {
            return getDefaultAttachmentSettings();
        } else {
            AttachmentSettings.Content attachmentSettingsContent
                    = BaseConfig.instance.gson.fromJson(settingsStr, AttachmentSettings.Content.class);
            if (attachmentSettingsContent == null) {
                return getDefaultAttachmentSettings();
            } else {
                return attachmentSettingsContent;
            }
        }
    }

    private AttachmentSettings.Content getDefaultAttachmentSettings() {
        return new AttachmentSettings.Content(30, new String[]{"jpeg", "jpg", "png", "pdf", "doc", "docx", "rtf"});
    }
}

File: threads/src/main/java/im/threads/ui/utils/FileHelper.kt
package im.threads.ui.utils

import android.annotation.SuppressLint
import im.threads.business.chat_updates.ChatUpdateProcessor
import im.threads.business.config.BaseConfig
import im.threads.business.logger.LoggerEdna
import im.threads.business.preferences.Preferences
import im.threads.business.serviceLocator.core.inject
import im.threads.business.transport.models.AttachmentSettings
import im.threads.ui.preferences.PreferencesUiKeys

/**
 * Вспомогательный класс для работы с файлами
 */
@SuppressLint("CheckResult")
object FileHelper {
    private val chatUpdateProcessor: ChatUpdateProcessor by inject()
    private val preferences: Preferences by inject()

    private const val MEGABYTE = (1024 * 1024).toDouble()

    init {
        chatUpdateProcessor.attachmentSettingsProcessor
            .subscribe(
                { receivedAttachmentSettings ->
                    attachmentSettings = receivedAttachmentSettings.content
                },
                LoggerEdna::error
            )
    }

    fun isAllowedFileSize(fileSize: Long): Boolean {
        return fileSize / MEGABYTE <= maxAllowedFileSize
    }

    fun isAllowedFileExtension(fileExtension: String?): Boolean {
        val attachmentSettings = attachmentSettings
        for (allowedExt in attachmentSettings.fileExtensions) {
            if (allowedExt.equals(fileExtension, ignoreCase = true)) {
                return true
            }
        }
        return false
    }

    fun isFileExtensionsEmpty(): Boolean {
        return attachmentSettings.fileExtensions.isNullOrEmpty()
    }

    fun isJpgAllow(): Boolean {
        return attachmentSettings.fileExtensions?.contains("jpg") == true
    }

    val maxAllowedFileSize: Long
        get() = attachmentSettings.maxSize.toLong()

    private var attachmentSettings: AttachmentSettings.Content
        get() {
            val settingsStr = preferences.get(PreferencesUiKeys.PREF_ATTACHMENT_SETTINGS) ?: ""

            return if (settingsStr.isEmpty()) {
                defaultAttachmentSettings
            } else {
                val attachmentSettingsContent =
                    BaseConfig.instance.gson.fromJson(settingsStr, AttachmentSettings.Content::class.java)
                attachmentSettingsContent ?: defaultAttachmentSettings
            }
        }
        set(value) = preferences.save(
            PreferencesUiKeys.PREF_ATTACHMENT_SETTINGS,
            BaseConfig.instance.gson.toJson(value)
        )

    private val defaultAttachmentSettings: AttachmentSettings.Content
        get() = AttachmentSettings.Content(30, arrayOf("jpeg", "jpg", "png", "pdf", "doc", "docx", "rtf"))
}
